src/hotspot/share/opto/addnode.cpp

```diff
@@ -395,9 +395,159 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
     }
   }
 
+  // Convert a + a + ... + a into a*n
+  Node* serial_additions = convert_serial_additions(phase, bt);
+  if (serial_additions != nullptr) {
+    return serial_additions;
+  }
+
   return AddNode::Ideal(phase, can_reshape);
 }
 
+// Try to convert a serial of additions into a single multiplication. Also convert `(a * CON) + a` to `(CON + 1) * a` as
```

Comment:
What about `(a * CON1) + (a * CON2)`? Like `11 * a + 5 * a`. Do we also optimize that?

Reply:
Ah interesting. It could be worth adding a comment for that here then!

Reply:
That was in fact a large part of my initial hesitation with this PR.

Reply:
Added comments on `(a*b) + (a*c)` and `AddNode::IdealIL`